File: cngx/src/main/java/impl/memory/Session.java
package impl.memory;

import api.memory.*;
import api.memory.elements.*;

import java.util.Map;
import java.util.UUID;

public class Session extends ISession {
	private Store store = null;
	private UUID uuid = UUID.randomUUID();
	private boolean isOpen = false;

	public Store getStore() {
		return store;
	}

	public void setStore(Store store) {
		this.store = store;
	}

	public UUID getUuid() {
		return uuid;
	}

	public void setUuid(UUID uuid) {
		this.uuid = uuid;
	}

	public boolean isOpen() {
		return isOpen;
	}

	public void setOpen(boolean isOpen) {
		this.isOpen = isOpen;
	}

	@Override
	public String createTag(String tag, Map<String, Value> properties) {
		return null;
	}

	@Override
	public String deleteTag(String tag) {
		return null;
	}

	@Override
	public String modifyTag(String tag, Map<String, String> properties) {
		return null;
	}

	@Override
	public ITag getTag(String tag) {
		return null;
	}

	@Override
	public INode createNode() {
		return null;
	}

	@Override
	public INode createNode(Map<String, Value> properties) {
		return null;
	}

	@Override
	public void deleteNode(INode node) {

	}

	@Override
	public INode modifyNode(Map<String, Value> properties) {
		return null;
	}

	@Override
	public INode[] searchNodes(String criteria) {
		return new INode[0];
	}

	@Override
	public INode getNode(String rid) {
		return null;
	}

	@Override
	public INode getNode(UUID uuid) {
		return null;
	};

	@Override
	public ILink createLink() {
		return null;
	}

	@Override
	public ILink createLink(Map<String, Value> properties) {
		return null;
	}

	@Override
	public ILink[] getOutLinks(INode node) {
		return new ILink[0];
	}

	@Override
	public ILink[] getInLinks(INode node) {
		return new ILink[0];
	}

	@Override
	public ILink[] getAllLinks(INode node) {
		return new ILink[0];
	}

	@Override
	public void connect(INode from, INode to, ILink link) {

	}

	@Override
	public void disconnect(INode from, INode to, ILink link) {

	}

	@Override
	public void disconnectInLinks(INode node) {

	}

	@Override
	public void disconnectOutLinks(INode node) {

	}

	@Override
	public void disconnectAlltLinks(INode node) {

	}



}

File: cngx/src/main/scala/org/compendiumng/cngx/memory/Session.scala
package org.compendiumng.cngx.memory

import java.io.File
import java.util.UUID
import com.typesafe.scalalogging.StrictLogging
import org.apache.commons.lang.NotImplementedException
import com.typesafe.config._


object Session extends StrictLogging {
  private var AllSessions = Map[String, Session]()

  val Config = ConfigFactory.load("cngx.properties")
  val cfgVersion = Config.getNumber("version")

  assert(Config.isResolved, "Configuration is not resolved")

  /**
   * @param props properties of the session
   * @return established session
   */
  def create(props: Map[String, Any]): Session = {
    logger.info("Creating session with following properties: " + props)


    val session_id: String = props("id").asInstanceOf[String]

    val s: Session = new Session(props, session_id)

    if (s.connect()) {
      logger.info("Sucessfully connected to: " + s)
      AllSessions += session_id -> s
      s
    } else {
      logger.error("Failed to connect to: " + s.toString)
      null
    }
  }
}

class Session (updateProperties: Map[String, Any],  session_id: String) {
  val config_id: Any = updateProperties.getOrElse("id", UUID.randomUUID().toString)
  val properties = Map[String, Any]()

  def createElement(props: Map[String, Any]): Element = {
		def e = new Element(props)
		e.properties ++  props
    e
	}

	def connect(): Boolean = {
		throw new NotImplementedException()
		false
	}

	def deleteElement(element: Element, uuid: UUID, id: String): Unit = {
		throw new NotImplementedException()
	}


  def updateElementProperties(element: Element, props: Map[String,Any]): Element = {
    throw new NotImplementedException()
    null
  }

	def searchNodes(criteria: String): Seq[Element] = {
    throw new NotImplementedException()
    null
	}

	def getNode(rid: String):Element = {
    throw new NotImplementedException()
    null
	}

	def getNode(uuid: UUID): Element = {
    throw new NotImplementedException()
	}

	def createLink(NodeA: Element, nodeB: Element, props: Map[String,Any]): Element = {
    throw new NotImplementedException()
		null
	}

	def getOutLinks(element: Element): Seq[Element] =  {
    throw new NotImplementedException()
    null
	}

	def getInLinks(element: Element): Seq[Element] =  {
    throw new NotImplementedException()
    null
	}

	def getAllLinks(element: Element): Seq[Element] =  {
    throw new NotImplementedException()
    null
	}

	def link(nodeA: Element, nodeB: Element, ILink: Element): Seq[Element] = {
    throw new NotImplementedException()
    null
	}

	def unlinkInLinks(node: Element): Seq[Element] = {
    throw new NotImplementedException()
    null
	}

	def unlinkOutLinks(node: Element): Seq[Element] =  {
    throw new NotImplementedException()
    null
	}

	def unlinkAllLinks(element: Element): Seq[Element] =  {
    throw new NotImplementedException()
    null
	}

  /**
   *  add a file as document with properties and link it to an Element
   * @param file file on filesystem
   * @param name identifier of the file within Element (must be unique within Element)
   * @return uuid of stored and document
   */
  def addDoc(file: File, name: String, props: Map[String, Any]): Option[UUID] = {
    var uuid: Option[UUID] = ???
    if (file.exists() && file.canRead) {
      // read it
      val docelement = createElement(
        Map("label" -> name, "elementType" -> "document",
          "orig_file" -> file.getName, "orig_path" -> file.getAbsolutePath))
      docelement.save()
      docelement.properties("uuid")

    } else {
      return new Option[null]
    }

  }

}
